test/bpf_tests/bpf_test.go

```diff
@@ -0,0 +1,443 @@
+		}
+
+		if !strings.HasSuffix(entry.Name(), ".o") {
+			continue
+		}
+
+		func() {
+			elfPath := path.Join(*testPath, entry.Name())
+			spec, err := ebpf.LoadCollectionSpec(elfPath)
+			if err != nil {
+				t.Fatal("load spec: ", elfPath, err)
+			}
+
+			for _, mspec := range spec.Maps {
+				// Drain extra info from map specs
+				if mspec.Extra != nil {
+					io.ReadAll(mspec.Extra)
+				}
+
+				// Remove BTF from maps as a workaround since some maps, like 'cilium_xdp_scratch' will return a
+				// "map create: invalid argument" error if we don't
+				mspec.BTF = nil
+			}
+
+			// Detect program type mismatches
+			var progTestType ebpf.ProgramType
+			for _, spec := range spec.Programs {
+				if progTestType != spec.Type {
+					if progTestType == ebpf.UnspecifiedProgram {
+						progTestType = spec.Type
+						continue
+					}
+					if spec.Type == ebpf.UnspecifiedProgram {
+						continue
+					}
+
+					t.Fatalf(
+						"File '%s' contains both '%s' and '%s' program types, "+
+							"only one program type per ELF file allowed:",
+						elfPath,
+						progTestType,
+						spec.Type,
+					)
+				}
+			}
+
+			if progTestType == ebpf.UnspecifiedProgram {
+				t.Fatalf("File '%s' only contains unspecified program types", elfPath)
+			}
+
+			// Give all tail call programs the same program type as the test programs
+			for _, spec := range spec.Programs {
+				spec.Type = progTestType
+			}
+
+			coll, err := ebpf.NewCollectionWithOptions(spec, ebpf.CollectionOptions{
+				Programs: ebpf.ProgramOptions{
+					LogSize: 1 << 20,
+				},
+				Maps: ebpf.MapOptions{
+					PinPath: pinPath,
+				},
+			})
+			if err != nil {
+				t.Fatal("new coll:", err)
+			}
+			defer coll.Close()
+
+			err = loadCallsMap(spec, coll)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			testNameToPrograms := make(map[string]programSet)
+
+			for progName, spec := range spec.Programs {
+				match := checkProgRegex.FindStringSubmatch(spec.SectionName)
+				if len(match) == 0 {
+					continue
+				}
+
+				progs := testNameToPrograms[match[1]]
+				if match[2] == "setup" {
+					progs.setupProg = coll.Programs[progName]
+				}
+				if match[2] == "check" {
+					progs.checkProg = coll.Programs[progName]
+				}
+				testNameToPrograms[match[1]] = progs
+			}
+
+			for progName, set := range testNameToPrograms {
+				if set.checkProg == nil {
+					t.Fatalf(
+						"File '%s' contains a setup program in section '%s' but no check program.",
+						elfPath,
+						spec.Programs[progName].SectionName,
+					)
+				}
+			}
+
+			// Collect debug events and add them as logs of the main test
+			var globalLogReader *perf.Reader
+			if coll.Maps["test_cilium_events"] != nil {
+				globalLogReader, err = perf.NewReader(coll.Maps["test_cilium_events"], 4096*16)
+				if err != nil {
+					t.Fatalf("new global log reader: %s", err.Error())
+				}
+
+				linkCache := link.NewLinkCache()
+
+				// Closes as soon as the goroutine has started
+				started := make(chan struct{})
```

Comment:
No, I don't think so. Since any messages accumulated should be buffered in the perf buffer after we called `perf.NewReader`. I think I was confused because `globalLogReader.Read` is blocking so I assumed the messages would get lost if there was not blocking goroutine ready. Will fix that.